### PR TITLE
Add evenly spaced streamlines 2D filter

### LINKED ISSUE #1360: 2d streamlines from a sliced 3D mesh data not working
**Plotting 2D streamlines**

My aim is to plot 2D streamlines on a plane, obtained upon slicing a 3D mesh. Though this is possible in Paraview, the 2D streamlines are very densely arranged. As the plots from pyvista look cool, I plan to get this working using pyvista itself. For this, I have actually exported a 2D mesh data (VTK format) from the 3D mesh by slicing (performed this in Paraview itself). Following the pyvista examples, using _mesh.streamlines()_ function gives me an empty array. Would appreciate if someone can give me clues to proceed.

-----

The colab notebook I have used for this is accessible at [Link1](https://colab.research.google.com/drive/1rMCW8CD4Nyq84zuRPExvt_AGqKMiHofn?usp=sharing) and the VTK file to try this at [Link2](https://drive.google.com/file/d/1_Lzzr937f2ABacnRsVzS8IhWf8ntIUWY/view?usp=sharing). The 2D contours are working (as seen in the colab notebook), so I am not sure why this is not the case with streamlines.

Also added a few screenshots to depict this problem : 

-----

**System Information:**

```txt
Date: Mon Jun 07 16:53:02 2021 UTC

                OS : Linux
            CPU(s) : 2
           Machine : x86_64
      Architecture : 64bit
               RAM : 12.7 GiB
       Environment : IPython
        GPU Vendor : VMware, Inc.
      GPU Renderer : llvmpipe (LLVM 10.0.0, 256 bits)
       GPU Version : 3.3 (Core Profile) Mesa 20.0.8

  Python 3.7.10 (default, May  3 2021, 02:48:31)  [GCC 7.5.0]

           pyvista : 0.30.1
               vtk : 9.0.1
             numpy : 1.19.5
           imageio : 2.4.1
           appdirs : 1.4.4
            scooby : 0.5.7
            meshio : 4.4.6
        matplotlib : 3.2.2
           IPython : 5.5.0
          colorcet : 1.0.0
             scipy : 1.4.1
              tqdm : 4.41.1
--------------------------------------------------------------------------------

```

### PR DESCRIPTION
**Pull request by MatthewFlamm**
### Overview

This PR enables usage of [vtkEvenlySpacedStreamlines2d](https://vtk.org/doc/nightly/html/classvtkEvenlySpacedStreamlines2D.html) as a filter `streamlines_evenly_spaced_2D`.  This only works on 2D datasets and provides a quick way to generate nice-ish streamline plots that cover the whole mesh.

This PR also adds an example of using streamlines on 2D datasets including the existing `streamlines` filter and the new  `streamlines_evenly_spaced_2D` filter.

Closes #1360 and related to https://github.com/pyvista/pyvista-support/issues/414

### Details

I couldn't find any good existing datasets to use for 2D streamlines.  There is one naca airfoil dataset, but it only includes density.  The antarctica dataset has 2D vectors, but the streamlines often follow the mesh edges instead of obtaining smooth curves.  I generated my own dataset, but it would have to be uploaded to vtk-data prior to fully testing this one.  I will paste the pictures from the added example here pending feedback on existing datasets or whether to add this dataset.

image
image
image

TODO:

- [x] Add dataset to vtk-data/or use a different dataset
- [x] Clean up example to reuse arguments between examples, e.g. compute_vorticity
- [x] Add tests
- [x] Clean up `streamlines_from_source` filter docstring to match changes in this filter
- [x] Add example to docstring

### CODE REVIEW

**Comment by tkoyama010** [comment]
This PR is awsome. Thanks @MatthewFlamm .

**Comment by MatthewFlamm** [comment]
I made https://github.com/pyvista/vtk-data/pull/2 to add the datafile used in this PR.

**Comment by MatthewFlamm** [comment]
Does the new example file need to get included somewhere in the doc building process?  I'm not sure, and I cannot build the documentation locally.

I also need help with this error, it seems unrelated, but again, I'm not sure.
https://github.com/pyvista/pyvista/pull/1441/checks?check_run_id=2953999908#step:6:839

**Comment by adeak** [comment]
Yeah, seems unrelated, and not even an error. If I understand correctly, there was a warning raised that says
```
vtkDebugLeaks has found no leaks.
```
which... is a good thing. Debug flags are enabled for the VTK 9.0.2 dev wheels, this is probably related. The issue is that warnings are automatically promoted to errors.

Paging @akaszynski for sage wisdom.

**Comment by akaszynski** [comment]
> Does the new example file need to get included somewhere in the doc building process? I'm not sure, and I cannot build the documentation locally.

No, you placed it in the correct location, so it should build (and in fact does):
```
    - ../examples/01-filter/streamlines_2D.py:                       4.21 sec   0.0 MB
```

**Comment by MatthewFlamm** [comment]
This PR would need to take into account #1478 or vice versa for the doctest.

File: examples/01-filter/streamlines_2D.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
+"""
```
Comment (akaszynski) [inline examples/01-filter/streamlines_2D.py:1]:
```suggestion
"""
.. _2d_streamlines_example:
```

File: pyvista/core/filters/data_set.py [hunk 2]
```diff
@@ -2173,6 +2173,160 @@ def streamlines_from_source(dataset, source, vectors=None,
...
+        >>> from pyvista import examples
+        >>> mesh = examples.download_cylinder_crossflow()
+        >>> streams = mesh[0].streamlines_evenly_spaced_2D(start_position=(4, 0.1, 0.),
+        ...                                                separating_distance=3,
+        ...                                                separating_distance_ratio=0.2)
+        >>> plotter = pyvista.Plotter()
+        >>> _ = plotter.add_mesh(streams.tube(radius=0.02), scalars="vorticity_mag")
+        >>> plotter.view_xy()
+        >>> _ = plotter.show()
+
```
Comment (akaszynski) [inline pyvista/core/filters/data_set.py:2273]:
```suggestion

        See :ref:`2d_streamlines_example` for more examples using this filter.
```

File: pyvista/core/filters/data_set.py [hunk 3]
```diff
@@ -2173,6 +2173,160 @@ def streamlines_from_source(dataset, source, vectors=None,
...
 
+    def streamlines_evenly_spaced_2D(dataset, vectors=None, start_position=None,
+                                    integrator_type=2, step_length=0.5, step_unit='cl',
+                                    max_steps=2000, terminal_speed=1e-12, interpolator_type='point',
+                                    separating_distance=10, separating_distance_ratio=None,
+                                    closed_loop_maximum_distance=0.5, loop_angle=20,  
+                                    minimum_number_of_loop_points=4, compute_vorticity=True):
+        """Generate evenly spaced streamlines on a 2D dataset.
+
+        This filter only supports datasets that lie on the xy plane, i.e. z=0.
```
Comment (akaszynski) [inline pyvista/core/filters/data_set.py:2184]:
```suggestion
        This filter only supports datasets that lie on the xy plane, i.e. ``z=0``.
```

File: pyvista/core/filters/data_set.py [hunk 4]
```diff
@@ -2173,6 +2173,160 @@ def streamlines_from_source(dataset, source, vectors=None,
...
+        Particular care must be used to choose a `separating_distance`
+        that do not result in too much memory being utilized.  The
+        default unit is cell length.
+
+        Parameters
+        ----------
+        vectors : str, optional
+            The string name of the active vector field to integrate across.
+
+        start_position : 3 element array
```
Comment (akaszynski) [inline pyvista/core/filters/data_set.py:2194]:
```suggestion
        start_position : sequence(float), optional
```

File: pyvista/core/filters/data_set.py [hunk 5]
```diff
@@ -2173,6 +2173,160 @@ def streamlines_from_source(dataset, source, vectors=None,
...
+        >>> import pyvista
+        >>> from pyvista import examples
+        >>> mesh = examples.download_cylinder_crossflow()
+        >>> streams = mesh[0].streamlines_evenly_spaced_2D(start_position=(4, 0.1, 0.),
+        ...                                                separating_distance=3,
+        ...                                                separating_distance_ratio=0.2)
+        >>> plotter = pyvista.Plotter()
+        >>> _ = plotter.add_mesh(streams.tube(radius=0.02), scalars="vorticity_mag")
+        >>> plotter.view_xy()
+        >>> _ = plotter.show()
```
Comment (akaszynski) [inline pyvista/core/filters/data_set.py:2272]:
```suggestion
        >>> plotter.show()
```
No longer necessary.

**Review by akaszynski — APPROVED** [review]
Approved with minor suggestions.  I've suggested we add a link in the examples section to the example in the gallery; we should do that with new examples from now.

**Pull request merged**